ecommerce/serializers.py

```diff
@@ -202,6 +204,24 @@ class ProductDetailSerializer(ProductSerializer):
 
     latest_version = serializers.SerializerMethodField()
 
+    @staticmethod
+    def setup_eager_loading(queryset):
+        """
+        setup django eager loading.
+        """
+        queryset = queryset.select_related("content_type")
+        courserun_ct = ContentType.objects.get_for_model(CourseRun)
+        tags_with_courserun = [item for item in queryset if item.content_type_id == courserun_ct.id]
+        prefetch_related_objects(tags_with_courserun, "content_object__course__courseruns")
```

Comment:
While testing locally this only seems to add more queries to the request. Perhaps because later on in the serializer the `course.unexpired_runs` is being called which applies a `filter` and `order_by` that gets rid of the prefetch cache altogether. Short of doing a rather invasive change I'd say the rest of the optimisations are fine. This however introduce an unnecessary eager load on `Course.courseruns`.

Reply:
Actually, the above code was only added for getting idea about the loading of `Course.courseruns` only. Yet the only optimization will be get through this only. 

```
    @staticmethod
    def setup_eager_loading(queryset):
        """
        setup django eager loading.
        """
        queryset = queryset.select_related("content_type")
        queryset = queryset.prefetch_related(
            dj_models.Prefetch(
                "productversions",
                queryset=models.ProductVersion.objects.order_by("-created_on"),
            )
        )        
        return queryset```